packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -114,22 +115,36 @@ class SocketReference extends TypedEventEmitter<ISocketEvents> {
         assert(!SocketReference.socketIoSockets.has(key), 0x220 /* "socket key collision" */);
         SocketReference.socketIoSockets.set(key, this);
 
-        // The server always closes the socket after sending this message
-        // fully remove the socket reference now
-        socket.on("server_disconnect", (socketError: IOdspSocketError) => {
+        const disconnectHandler = (socketError: IOdspSocketError, clientId?: string) => {
             // Treat all errors as recoverable, and rely on joinSession / reconnection flow to
             // filter out retryable vs. non-retryable cases.
-            const error = errorObjectFromSocketError(socketError, "server_disconnect");
+            const error = errorObjectFromSocketError(socketError, clientId
+                ? "disconnect_document" : "server_disconnect");
             error.canRetry = true;
 
             // see comment in disconnected() getter
             // Setting it here to ensure socket reuse does not happen if new request to connect
             // comes in from "disconnect" listener below, before we close socket.
             this.isPendingInitialConnection = false;
 
-            // Explicitly cast error to the specified event args type to ensure type compatibility
+            if (clientId !== undefined) {
+                // We don't need to close the socket in case of this event.
+                this.emit("disconnect_document", error, clientId);
+                return;
+            }
             this.emit("server_disconnect", error);
```

Comment:
I'd suggest changing event names / flow. disconnect_document / server_disconnect are name of protocol messages, we processed them, and now this layer tells consumers to disconnect - either specific client or socket. I'd either have appropriate names (something like "disconnectClient", "disconnectSocket"), or have just one event for both (surprisingly, we do not use "disconnect" on SocketReference) and just treat clientId === undefined as indication that all clients need to be disconnected.

BTW, why not use same on the wire? I.e. why not add extra arg (clientId) to existing server_disconnect and have just one type of message?

Reply:
I'd think we do not need supportedFeatures = { server_disconnect_document : true } with that proposal.
Clients who do not understand new arg would drop whole socket - not efficient, but it's correct, and I assume that's what happens today anyway (i.e. PUSH has no other tool to use). This (lack of new features and different code paths on PUSH) will make it simpler; it seems.

Reply:
That said, take a look at https://dev.azure.com/fluidframework/internal/_workitems/edit/2013 I've just created. I sort of contradict myself here, in a sense of asking for full socket closure to go on its own path. Curious what you think

Reply:
I like that idea about reusing server_disconnect. Having the server know about the server_disconnect_document feature was helpful because it let Push know whether it has to disconnect the websocket or if it could just emit a disconnect_document for a client. It does seem like adding a client id arg to server_disconnect could let us not need to know this.